ProcessLib/ComponentTransport/ComponentTransportFEM.h

```diff
@@ -131,48 +152,119 @@ class LocalAssemblerData : public ComponentTransportLocalAssemblerInterface
         auto local_b = MathLib::createZeroedVector<LocalVectorType>(
             local_b_data, local_matrix_size);
 
+        // Nodal DOFs include pressure
+        auto const num_comp = num_nodal_dof - 1;
+
+        // Get block matrices
+        auto Kpp = local_K.template block<pressure_size, pressure_size>(
+            pressure_index, pressure_index);
+        auto Mpp = local_M.template block<pressure_size, pressure_size>(
+            pressure_index, pressure_index);
+        auto Bp = local_b.template segment<pressure_size>(pressure_index);
+
+        auto local_p = Eigen::Map<const NodalVectorType>(
+            &local_x[pressure_index], pressure_size);
+
+        for (int comp_id = 0; comp_id < num_comp; ++comp_id)
+        {
+            /*  Partitioned assembler matrix
+             *  |  pp | pc1 | pc2 | pc3 |
+             *  |-----|-----|-----|-----|
+             *  | c1p | c1c1|  0  |  0  |
+             *  |-----|-----|-----|-----|
+             *  | c2p |  0  | c2c2|  0  |
+             *  |-----|-----|-----|-----|
+             *  | c3p |  0  |  0  | c3c3|
+             */
+            auto concentration_index =
+                pressure_size + comp_id * concentration_size;
+
+            auto KCC =
+                local_K.template block<concentration_size, concentration_size>(
+                    concentration_index, concentration_index);
+            auto MCC =
+                local_M.template block<concentration_size, concentration_size>(
+                    concentration_index, concentration_index);
+            auto MCp =
+                local_M.template block<concentration_size, pressure_size>(
+                    concentration_index, pressure_index);
+            auto MpC =
+                local_M.template block<pressure_size, concentration_size>(
+                    pressure_index, concentration_index);
+
+            auto local_C = Eigen::Map<const NodalVectorType>(
+                &local_x[concentration_index], concentration_size);
+
+            // Prevent duplicate computation in loops
```

Comment:
What does this mean? As far as I can see the pure pressure parts, i.e., Mpp and Kpp are recomputed for each component. Is this avoidable.

Reply:
Fixed.